### PR TITLE
Add more Unit tests

### PR DESCRIPTION
**Pull request by nilasini**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+        registrationRequestFactory.create(mockRegistrationRequestBuilder, mockHttpRequest, mockHttpResponse);
+
+        if (clientName != null) {
+            assertEquals(registrationRequestProfile.getClientName(), clientName);
+        }
+
+        if (!expected.equals("empty")) {
+            if (expected instanceof String) {
+                assertEquals(registrationRequestProfile.getGrantTypes().get(0), grantType);
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java]: Please add messages.

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        JSONArray responseTypes = new JSONArray();
+        JSONArray scopes = new JSONArray();
+        JSONArray contacts = new JSONArray();
+        grantTypes.add("dummyGrantType");
+        redirectUrls.add("dummyRedirectUrl");
+        responseTypes.add("dummyResponseType");
+        contacts.add("dummyContact");
+        scopes.add("dummyScope");
+
+        String grantType = "dummyGrantType";
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java]: Why you define constant here and use plain string in line [81](https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/574/files#diff-ea882f3939d7b33008446ae944e7fa5fR81) ?

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
+package org.wso2.carbon.identity.oauth.dcr.factory;
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java]: Insert license.

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+                assertEquals(registrationRequestProfile.getRedirectUris(), redirectUrl,
+                        "expected redirect url is not found in registrationRequestProfile");
+                assertEquals(registrationRequestProfile.getContacts(), contact,
+                        "expected contact is not found in registrationRequestProfile");
+                assertEquals(registrationRequestProfile.getScopes(), scope,
+                        "expected scope is not found in registrationRequestProfile");
+                assertEquals(registrationRequestProfile.getResponseTypes(), responseType,
+                        "expected response type is not found in registrationRequestProfile");
+            }
+
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java]: Remove unnecessary new line

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+                MediaType.APPLICATION_JSON);
+        assertEquals((int) statusCode[0], HttpServletResponse.SC_BAD_REQUEST, "Status code doesn't match with "
+                + HttpServletResponse.SC_BAD_REQUEST);
+    }
+
+    @Test
+    public void testGenerateErrorResponse() throws Exception {
+
+        String dummyError = "dummyError";
+
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java]: Remove unnecessary new line.

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+        suppress(methodsDeclaredIn(HttpIdentityRequestFactory.class));
+
+        when(mockHttpRequest.getReader()).thenReturn(mockReader);
+        whenNew(JSONParser.class).withNoArguments().thenReturn(jsonParser);
+
+        when(jsonParser.parse(mockReader)).thenReturn(jsonObject);
+
+        String carbonHome = Paths.get(System.getProperty("user.dir"), "src", "test", "resources").toString();
+        System.setProperty(CarbonBaseConstants.CARBON_HOME, carbonHome);
+        PrivilegedCarbonContext.startTenantFlow();
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java]: Do you end the tenant flow ?

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java (hunk 3 above)
Comment (nilasini) [reply components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java]: inserted

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+                        "expected contact is not found in registrationRequestProfile");
+                assertEquals(registrationRequestProfile.getScopes(), scope,
+                        "expected scope is not found in registrationRequestProfile");
+                assertEquals(registrationRequestProfile.getResponseTypes(), responseType,
+                        "expected response type is not found in registrationRequestProfile");
+            }
+        }
+        assertEquals(registrationRequestProfile.getOwner(), ownerName,
+                "expected owner name is not found in registrationRequestProfile");
+        PrivilegedCarbonContext.endTenantFlow();
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java]: use a try finally block to end tenant flow.

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+                header[0] = (String) invocation.getArguments()[1];
+                return null;
+            }
+        }).when(mockHttpIdentityResponseBuilder).addHeader(anyString(), anyString());
+
+        FrameworkClientException exception = mock(FrameworkClientException.class);
+        when(exception.getMessage()).thenReturn(dummyDescription);
+        registrationRequestFactory.handleException(exception, mockHttpRequest, mockHttpResponse);
+
+        assertEquals(header[0], MediaType.APPLICATION_JSON, "Header doesn't match with " +
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java]: Can you guarantee that the first header will be this ? I had experience that things come in different order.

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+        doAnswer(new Answer<Object>() {
+            @Override
+            public Object answer(InvocationOnMock invocation) throws Throwable {
+
+                header[0] = (String) invocation.getArguments()[1];
+                return null;
+            }
+        }).when(mockHttpIdentityResponseBuilder).addHeader(eq(OAuthConstants.HTTP_RESP_HEADER_CACHE_CONTROL),
+                anyString
+                        ());
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java]: formatting

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+        } finally {
+            PrivilegedCarbonContext.endTenantFlow();
+        }
+    }
+
+    @Test(expectedExceptions = IdentityException.class)
+    public void testCreateWithEmptyRedirectUri() throws Exception {
+
+        String grantType = "implicit";
+        int redirectUrl = 0;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactoryTest.java]: redirectUrl is an int? :open_mouth:

**Pull request merged**